crates/services/compression/src/service.rs

```diff
@@ -1,69 +1,146 @@
-use crate::ports::{
-    block_source::BlockSource,
-    compression_storage::CompressionStorage,
-    configuration::CompressionConfigProvider,
+use crate::{
+    config::CompressionConfig,
+    ports::{
+        block_source::BlockSource,
+        compression_storage::{
+            CompressionStorage,
+            HistoricalCompressionStorage,
+            WriteCompressedBlock,
+        },
+        configuration::CompressionConfigProvider,
+    },
+    temporal_registry::{
+        CompressionContext,
+        CompressionStorageWrapper,
+    },
 };
+use fuel_core_compression::compress::compress;
 use fuel_core_services::{
-    EmptyShared,
     RunnableService,
     RunnableTask,
+    SeqLock,
+    SeqLockReader,
+    SeqLockWriter,
     ServiceRunner,
     StateWatcher,
 };
+use futures::FutureExt;
+
+/// Shared state for the compression service.
+#[derive(Debug, Clone)]
+pub struct SharedState {
+    #[allow(unused)] // just for now
+    latest_height: SeqLockReader<Option<u32>>,
+}
 
 /// The compression service.
 /// Responsible for subscribing to the l2 block stream,
 /// perform compression of the block, and store the compressed block.
 /// We don't need to share any data between this task and anything else yet(?)
 /// perhaps we want to expose a way to get the registry root
-pub struct CompressionService<B, S, C> {
-    /// The block source.
-    block_source: B,
+pub struct CompressionService<S> {
+    /// The block stream.
+    block_stream: crate::ports::block_source::BlockStream,
     /// The compression storage.
     storage: S,
     /// The compression config.
-    config: C,
+    config: CompressionConfig,
+    /// shared state
+    shared_state: SharedState,
+    /// writer of shared state
+    shared_state_writer: SeqLockWriter<Option<u32>>,
 }
 
-impl<B, S, C> CompressionService<B, S, C>
+use fuel_core_storage::transactional::WriteTransaction;
+
+impl<S> CompressionService<S>
 where
-    B: BlockSource,
-    S: CompressionStorage,
-    C: CompressionConfigProvider,
+    S: CompressionStorage + HistoricalCompressionStorage,
 {
-    fn new(block_source: B, storage: S, config: C) -> Self {
+    fn new(
+        block_stream: crate::ports::block_source::BlockStream,
+        storage: S,
+        config: CompressionConfig,
+    ) -> Self {
+        let latest_height = storage.latest_height();
+
+        // SAFETY: we are creating a new SeqLock with a valid value
+        let (writer, reader) = unsafe { SeqLock::new(latest_height) };
```

Comment:
It would be nice to skip the seqlock. Do we really need it?

Reply:
seqlock's are best for small data, which this is ~ we made it explicitly unsafe to create one because then it enforces the implementer to actually see why it's unsafe. i like it :)